File: ThalesCore/Cryptography/TripleDES.vb
''
'' This program is free software; you can redistribute it and/or modify
'' it under the terms of the GNU General Public License as published by
'' the Free Software Foundation; either version 2 of the License, or
'' (at your option) any later version.
''
'' This program is distributed in the hope that it will be useful,
'' but WITHOUT ANY WARRANTY; without even the implied warranty of
'' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'' GNU General Public License for more details.
''
'' You should have received a copy of the GNU General Public License
'' along with this program; if not, write to the Free Software
'' Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
'' 

Namespace Cryptography

    ''' <summary>
    ''' Utility class to perform Triple DES operations.
    ''' </summary>
    ''' <remarks>
    ''' This class can be used to perform 3D operations using <see cref="HexKey"/> hexadecimal keys.
    ''' </remarks>
    Public Class TripleDES

        ''' <summary>
        ''' Performs an encryption operation.
        ''' </summary>
        ''' <remarks>
        ''' Performs an encryption operation.
        ''' </remarks>
        Public Shared Function TripleDESEncrypt(ByVal key As HexKey, ByVal data As String) As String

            If data Is Nothing AndAlso data.Length <> 16 AndAlso data.Length <> 32 AndAlso data.Length <> 48 Then Throw New Exceptions.XInvalidData("Invalid data for 3DEncrypt with variant")

            Dim result As String
            If data.Length = 16 Then
                result = TripleDESEncryptSingleLength(key, data)
            ElseIf data.Length = 32 Then
                result = TripleDESEncryptSingleLength(key, data.Substring(0, 16)) + _
                         TripleDESEncryptSingleLength(key, data.Substring(16, 16))
            Else
                result = TripleDESEncryptSingleLength(key, data.Substring(0, 16)) + _
                         TripleDESEncryptSingleLength(key, data.Substring(16, 16)) + _
                         TripleDESEncryptSingleLength(key, data.Substring(32, 16))
            End If
            Return result
        End Function

        Private Shared Function TripleDESEncryptSingleLength(ByVal key As HexKey, ByVal data As String) As String
            Dim result As String = ""
            result = DES.DESEncrypt(key.PartA, data)
            result = DES.DESDecrypt(key.PartB, result)
            result = DES.DESEncrypt(key.PartC, result)
            Return result
        End Function

        ''' <summary>
        ''' Performs a decrypt operation.
        ''' </summary>
        ''' <remarks>
        ''' Performs a decryption operation.
        ''' </remarks>
        Public Shared Function TripleDESDecrypt(ByVal key As HexKey, ByVal data As String) As String
            If data Is Nothing AndAlso data.Length <> 16 AndAlso data.Length <> 32 AndAlso data.Length <> 48 Then Throw New Exceptions.XInvalidData("Invalid data for 3DEncrypt with variant")

            Dim result As String
            If data.Length = 16 Then
                result = TripleDESDecryptSingleLength(key, data)
            ElseIf data.Length = 32 Then
                result = TripleDESDecryptSingleLength(key, data.Substring(0, 16)) + _
                         TripleDESDecryptSingleLength(key, data.Substring(16, 16))
            Else
                result = TripleDESDecryptSingleLength(key, data.Substring(0, 16)) + _
                         TripleDESDecryptSingleLength(key, data.Substring(16, 16)) + _
                         TripleDESDecryptSingleLength(key, data.Substring(32, 16))
            End If
            Return result
        End Function

        Private Shared Function TripleDESDecryptSingleLength(ByVal key As HexKey, ByVal data As String) As String
            Dim result As String
            result = DES.DESDecrypt(key.PartC, data)
            result = DES.DESEncrypt(key.PartB, result)
            result = DES.DESDecrypt(key.PartA, result)
            Return result
        End Function

        ''' <summary>
        ''' Performs an encrypt operation with an applied variant.
        ''' </summary>
        ''' <remarks>
        ''' Performs an encrypt operation with an applied variant.
        ''' </remarks>
        Public Shared Function TripleDESEncryptVariant(ByVal key As HexKey, ByVal data As String) As String

            If data Is Nothing AndAlso data.Length <> 32 AndAlso data.Length <> 48 Then Throw New Exceptions.XInvalidData("Invalid data for 3DEncrypt with variant")

            Dim result1 As String, result2 As String, result3 As String = "", orgKeyPartB As String
            If data.Length = 32 Then
                orgKeyPartB = key.PartB
                key.PartB = Core.Utility.XORHexStrings(key.PartB, Cryptography.LMK.Variants.DoubleLengthVariant(1).PadRight(16, "0"c))
                result1 = TripleDESEncrypt(key, data.Substring(0, 16))
                key.PartB = Core.Utility.XORHexStringsFull(orgKeyPartB, Cryptography.LMK.Variants.DoubleLengthVariant(2).PadRight(16, "0"c))
                result2 = TripleDESEncrypt(key, data.Substring(16, 16))
            Else
                orgKeyPartB = key.PartB
                key.PartB = Core.Utility.XORHexStringsFull(key.PartB, Cryptography.LMK.Variants.TripleLengthVariant(1).PadRight(16, "0"c))
                result1 = TripleDESEncrypt(key, data.Substring(0, 16))
                key.PartB = Core.Utility.XORHexStringsFull(orgKeyPartB, Cryptography.LMK.Variants.TripleLengthVariant(2).PadRight(16, "0"c))
                result2 = TripleDESEncrypt(key, data.Substring(16, 16))
                key.PartB = Core.Utility.XORHexStringsFull(orgKeyPartB, Cryptography.LMK.Variants.TripleLengthVariant(3).PadRight(16, "0"c))
                result3 = TripleDESEncrypt(key, data.Substring(32, 16))
            End If
            Return result1 + result2 + result3
        End Function

        ''' <summary>
        ''' Performs a decrypt operation with an applied variant.
        ''' </summary>
        ''' <remarks>
        ''' Performs a decrypt operation with an applied variant.
        ''' </remarks>
        Public Shared Function TripleDESDecryptVariant(ByVal key As HexKey, ByVal data As String) As String

            If data Is Nothing AndAlso data.Length <> 32 AndAlso data.Length <> 48 Then Throw New Exceptions.XInvalidData("Invalid data for 3DDecrypt with variant")

            Dim result1 As String = "", result2 As String = "", result3 As String = "", orgKeyPartB As String
            If data.Length = 32 Then
                orgKeyPartB = key.PartB
                key.PartB = Core.Utility.XORHexStrings(key.PartB, Cryptography.LMK.Variants.DoubleLengthVariant(1).PadRight(16, "0"c))
                result1 = TripleDESDecrypt(key, data.Substring(0, 16))
                key.PartB = Core.Utility.XORHexStringsFull(orgKeyPartB, Cryptography.LMK.Variants.DoubleLengthVariant(2).PadRight(16, "0"c))
                result2 = TripleDESDecrypt(key, data.Substring(16, 16))
            ElseIf data.Length = 48 Then
                orgKeyPartB = key.PartB
                key.PartB = Core.Utility.XORHexStringsFull(key.PartB, Cryptography.LMK.Variants.TripleLengthVariant(1).PadRight(16, "0"c))
                result1 = TripleDESDecrypt(key, data.Substring(0, 16))
                key.PartB = Core.Utility.XORHexStringsFull(orgKeyPartB, Cryptography.LMK.Variants.TripleLengthVariant(2).PadRight(16, "0"c))
                result2 = TripleDESDecrypt(key, data.Substring(16, 16))
                key.PartB = Core.Utility.XORHexStringsFull(orgKeyPartB, Cryptography.LMK.Variants.TripleLengthVariant(3).PadRight(16, "0"c))
                result3 = TripleDESDecrypt(key, data.Substring(32, 16))
            End If

            Return result1 + result2 + result3

        End Function

    End Class

End Namespace

File: ThalesSim.Core/Cryptography/DES/TripleDes.cs
/*
 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/ 

using System;
using System.IO;
using System.Reflection;
using System.Security.Cryptography;
using ThalesSim.Core.Utility;
using log4net;

namespace ThalesSim.Core.Cryptography.DES
{
    public class TripleDes
    {
        public static byte[] DesEncrypt (byte[] key, byte[] data)
        {
            var log = LogManager.GetLogger("DesEncrypt");

            try
            {
                return DesOperation(key, data, true, log);
            }
            catch (Exception ex)
            {
                log.Error("DESEncrypt error", ex);
                throw;
            }
        }

        public static byte[] DesDecrypt(byte[] key, byte[] data)
        {
            var log = LogManager.GetLogger("DesDecrypt");

            try
            {
                return DesOperation(key, data, false, log);
            }
            catch (Exception ex)
            {
                log.Error("DESEncrypt error", ex);
                throw;
            }
        }

        public static string DesEncrypt (string key, string data)
        {
            return DesEncrypt(key.GetHexBytes(), data.GetHexBytes()).GetHexString();
        }

        public static string DesDecrypt (string key, string data)
        {
            return DesDecrypt(key.GetHexBytes(), data.GetHexBytes()).GetHexString();
        }

        public static byte[] TripleDesEncrypt (byte[] key1, byte[] key2, byte[] key3, byte[] data)
        {
            var result = DesEncrypt(key1, data);
            var result2 = DesDecrypt(key2, result);
            return DesEncrypt(key3, result2);
        }

        public static byte[] TripleDesDecrypt(byte[] key1, byte[] key2, byte[] key3, byte[] data)
        {
            var result = DesDecrypt(key3, data);
            result = DesEncrypt(key2, result);
            return DesDecrypt(key1, result);
        }

        public static string TripleDesEncrypt (string key1, string key2, string key3, string data)
        {
            return TripleDesEncrypt(key1.GetHexBytes(), key2.GetHexBytes(), key3.GetHexBytes(), data.GetHexBytes()).GetHexString();
        }

        public static string TripleDesDecrypt(string key1, string key2, string key3, string data)
        {
            return TripleDesDecrypt(key1.GetHexBytes(), key2.GetHexBytes(), key3.GetHexBytes(), data.GetHexBytes()).GetHexString();
        }

        private static byte[] DesOperation (byte[] key, byte[] data, bool encrypt, ILog log)
        {
            if (key == null || data == null)
            {
                throw new InvalidOperationException("Key or data cannot be null");
            }

            if (key.Length != 8 || data.Length != 8)
            {
                throw new InvalidOperationException("Key or data not eight bytes");
            }

            using (var desAlg = System.Security.Cryptography.DES.Create())
            {
                var nullVector = new byte[] { 0, 0, 0, 0, 0, 0, 0, 0 };
                var result = new byte[8];

                if (System.Security.Cryptography.DES.IsWeakKey(key) || System.Security.Cryptography.DES.IsSemiWeakKey(key))
                {
                    log.Warn(encrypt
                                 ? "***DES encrypt with weak or semi-weak key"
                                 : "***DES decrypt with weak or semi-weak key");
                }

                desAlg.Mode = CipherMode.ECB;
                desAlg.IV = nullVector;
                desAlg.Padding = PaddingMode.None;

                ICryptoTransform desTransform;

                var mi = desAlg.GetType().GetMethod("_NewEncryptor", BindingFlags.NonPublic | BindingFlags.Instance);
                try
                {
                    desTransform = (ICryptoTransform)mi.Invoke(desAlg, new object[] { key, desAlg.Mode, nullVector, desAlg.FeedbackSize, encrypt ? 0 : 1 });
                }
                catch (Exception ex)
                {
                    log.Warn("Transform invocation error", ex);
                    desTransform = encrypt ? desAlg.CreateEncryptor(key, nullVector) : desAlg.CreateDecryptor(key, nullVector);
                }

                using (var outStream = new MemoryStream(result))
                {
                    using (var cryptoStream = new CryptoStream(outStream, desTransform, CryptoStreamMode.Write))
                    {
                        cryptoStream.Write(data, 0, 8);
                        cryptoStream.Close();
                    }
                }

                return result;
            }
        }
    }
}
